File: app/src/main/java/com/bapplications/maplemobile/gameplay/GameMap.java
package com.bapplications.maplemobile.gameplay;

import android.util.Log;

import com.bapplications.maplemobile.gameplay.map.map_objects.MapCharacters;
import com.bapplications.maplemobile.input.EventsQueue;
import com.bapplications.maplemobile.input.events.Event;
import com.bapplications.maplemobile.input.events.EventListener;
import com.bapplications.maplemobile.input.events.EventType;
import com.bapplications.maplemobile.input.events.ItemDroppedEvent;
import com.bapplications.maplemobile.utils.Rectangle;
import com.bapplications.maplemobile.utils.StaticUtils;
import com.bapplications.maplemobile.constatns.Loaded;
import com.bapplications.maplemobile.gameplay.audio.Sound;
import com.bapplications.maplemobile.gameplay.map.map_objects.Drop;
import com.bapplications.maplemobile.gameplay.map.map_objects.DropSpawn;
import com.bapplications.maplemobile.gameplay.map.Layer;
import com.bapplications.maplemobile.gameplay.map.map_objects.MapBackgrounds;
import com.bapplications.maplemobile.gameplay.map.map_objects.MapDrops;
import com.bapplications.maplemobile.gameplay.map.MapInfo;
import com.bapplications.maplemobile.gameplay.map.map_objects.MapMobs;
import com.bapplications.maplemobile.gameplay.map.map_objects.MapPortals;
import com.bapplications.maplemobile.gameplay.map.map_objects.MapTilesObjs;
import com.bapplications.maplemobile.gameplay.audio.Music;
import com.bapplications.maplemobile.gameplay.map.Portal;
import com.bapplications.maplemobile.gameplay.map.map_objects.mobs.Attack;
import com.bapplications.maplemobile.gameplay.map.map_objects.mobs.MobSpawn;
import com.bapplications.maplemobile.gameplay.physics.Physics;
import com.bapplications.maplemobile.gameplay.player.Player;
import com.bapplications.maplemobile.utils.Point;
import com.bapplications.maplemobile.pkgnx.NXNode;
import com.bapplications.maplemobile.input.InputAction;

import org.jetbrains.annotations.NotNull;

public class GameMap implements EventListener {


    private int mapid;
    private State state;
    private MapMobs mobs;
    private Player player;
    private Camera camera;
    private MapDrops drops;
    private MapInfo mapInfo;
    private Physics physics;
    private MapPortals portals;
    private MapTilesObjs tilesobjs;
    private MapCharacters characters;
    private MapBackgrounds backgrounds;

    public enum State
    {
        INACTIVE,
        TRANSITION,
        ACTIVE
    };
    
    public GameMap(Camera camera)
    {
        state = State.INACTIVE;
        this.camera = camera;
        EventsQueue.Companion.getInstance().registerListener(EventType.ItemDropped, this);
    }

    public void init(int mapid){
        // drops.init();
        loadMap(mapid);
    }

    @Override
    public void onEventReceive(@NotNull Event event) {
        switch (event.getType()) {
            case ItemDropped:
                ItemDroppedEvent _event = (ItemDroppedEvent) event;
                if(mapid == _event.getMapId()) {
                    spawnItemDrop(_event.getOid(), _event.getId(), _event.getStart(), _event.getOwner());
                }
                break;
        }
    }

    public void spawnMobs(NXNode src) {
        int oid = 100; // todo: needs a way to calculate that
        for(NXNode spawnNode : src) {
            if(!spawnNode.getChild("type").get("").equals("m")) {
                continue;
            }
            String id = spawnNode.getChild("id").get("");
            short flip = spawnNode.getChild("f").get(0L).shortValue();
            Point p = new Point(spawnNode);
            MobSpawn spawn = new MobSpawn(oid++, Integer.parseInt(id), (byte)0, (byte)0, flip, true, (byte)0, p.flipY());
            mobs.spawn(spawn);
        }
    }

    public void spawnItemDrop(int oid, int id, Point start, int owner) {
        DropSpawn spawn = new DropSpawn(oid, id, id == 0, owner, start, Drop.State.DROPPED, true );
        drops.spawn(spawn);
    }

    void loadMap(int mapid)
    {
        this.mapid = mapid;
        String strid = StaticUtils.extendId(mapid, 9);
        char prefix = strid.charAt(0);

        NXNode src = mapid == -1 ? null : Loaded.getFile(Loaded.WzFileName.MAP).getRoot().getChild("Map").getChild("Map" + prefix).getChild(strid + ".img");

        // in case of no map exist with this mapid
        if (src != null && !src.isNotExist()) {
            try {
                mobs = new MapMobs();
                drops = new MapDrops();
                characters = new MapCharacters();
                physics = new Physics(src.getChild("foothold"));
                backgrounds = new MapBackgrounds(src.getChild("back"));
                portals = new MapPortals(src.getChild("portal"), mapid);

                mapInfo = new MapInfo(src, physics.getFHT().getWalls(), physics.getFHT().getBorders());
                tilesobjs = new MapTilesObjs(src, new Rectangle(mapInfo.getWalls(), mapInfo.getBorders()));
                spawnMobs(src.getChild("life"));
            } catch (Exception e) {
                Log.e("GameMap", "Error loading map " + mapid);
                throw e;
            }
        }
    }


    public void respawn(Point position)
    {
        Music.play(mapInfo.getBgm());

//        Point<int16_t> spawnpoint = portals.get_portal_by_id(portalid);
        Point startpos = physics.getYBelow(position);
        player.respawn(startpos, mapInfo.isUnderwater());
        camera.setView(mapInfo.getWalls(), mapInfo.getBorders());
        camera.update(player.getPosition().negateSign());
    }

    public void update(int deltatime) {
        if (state != State.ACTIVE)
            return;

////        combat.update();
        backgrounds.update(deltatime);
////        effect.update();
        tilesobjs.update(deltatime);
//
////        reactors.update(physics);
////        npcs.update(physics);
        mobs.update(physics, deltatime);
        characters.update(physics, deltatime);
        drops.update(physics, deltatime);
        player.update(physics, deltatime);
        portals.update(player.getPosition(), deltatime);
        camera.update(player.getPosition());

        if (!player.isClimbing()/* && !player.is_sitting()*/ && !player.isAttacking())
        {
            if (player.isPressed(InputAction.UP_ARROW_KEY) && !player.isPressed(InputAction.DOWN_ARROW_KEY))
                checkLadders(true);

            if (player.isPressed(InputAction.DOWN_ARROW_KEY))
                checkLadders(false);

            if (player.isPressed(InputAction.UP_ARROW_KEY))
                checkPortals();


//            if (player.isPressed(InputAction.SIT))
//            check_seats();

//            if (player.isPressed(InputAction.ATTACK))
//            combat.use_move(0);
//
//            if (player.isPressed(InputAction.PICKUP))
//            check_drops();
        }


        if (player.isInvincible())
            return;

        int oid = mobs.findColliding(player);
        if (oid != 0)
        {
            Attack.MobAttack attack = mobs.createAttack(oid);
            if (attack.isValid())
            {
                Attack.MobAttackResult result = player.damage(attack);
            }
        }

    }

    public Portal getPortalByName(String portalName) {
        return portals.getPortalByName(portalName);
    }

    private void checkPortals() {
        if (player.isAttacking())
            return;

        Point playerpos = player.getPosition();
        Portal.WarpInfo warpinfo = portals.findWarpAt(playerpos);

        if (warpinfo.intramap)
        {
            Portal spawnpoint = portals.getPortalByName(warpinfo.toname);
            Point startpos = physics.getYBelow(spawnpoint.getSpawnPosition());

            player.respawn(startpos, mapInfo.isUnderwater());
        }
        else if (warpinfo.valid)
        {
            (new Sound(Sound.Name.PORTAL)).play();
            GameEngine.getInstance().changeMap(warpinfo.mapid, warpinfo.toname);
        }
    }

    public void draw(float alpha)
    {
        if (state != State.ACTIVE)
            return;

        Point viewpos = camera.position(alpha);

        backgrounds.drawBackgrounds(viewpos, alpha);

        for (Layer id : Layer.values())
        {
            tilesobjs.draw(id, camera.getPositionRect(), viewpos, alpha);
            physics.draw(viewpos);
//            reactors.draw(id, viewx, viewy, alpha);
//            npcs.draw(id, viewx, viewy, alpha);
            mobs.draw(id, viewpos, alpha);
            characters.draw(id, viewpos, alpha);
            player.draw(id, viewpos, alpha);
            drops.draw(id, viewpos, alpha);
        }
//
//        combat.draw(viewx, viewy, alpha);
        portals.draw(viewpos, alpha);
        backgrounds.drawForegrounds(viewpos, alpha);
//        effect.draw();
        // for debugging purposes
//        camera.draw();

    }


    public void clear()
    {
        state = State.INACTIVE;
//        Texture.clear();
//        chars.clear();
//        npcs.clear();
//        mobs.clear();
//        drops.clear();
//        reactors.clear();
    }

    public int getMapId() {
        return mapid;
    }

    public Camera getCamera() {
        return camera;
    }

    public Player getPlayer() {
        return player;
    }

    public State getState() {
        return state;
    }

    public void checkLadders(boolean up)
    {
        if (!player.canClimb() || player.isClimbing() || player.isAttacking())
            return;

        player.setLadder(mapInfo.findLadder(player.getPosition(), up));
    }


    public void enterMap(Player player, Portal portal) {
        this.player = player;
        player.setMap(this);
        state = State.ACTIVE;
        respawn(portal.getSpawnPosition());
    }

}

File: app/src/main/java/com/bapplications/maplemobile/gameplay/GameMap.kt
package com.bapplications.maplemobile.gameplay

import android.util.Log
import com.bapplications.maplemobile.constatns.Loaded
import com.bapplications.maplemobile.gameplay.audio.Music
import com.bapplications.maplemobile.gameplay.audio.Sound
import com.bapplications.maplemobile.gameplay.map.Layer
import com.bapplications.maplemobile.gameplay.map.MapInfo
import com.bapplications.maplemobile.gameplay.map.MapObject
import com.bapplications.maplemobile.gameplay.map.Portal
import com.bapplications.maplemobile.gameplay.map.map_objects.*
import com.bapplications.maplemobile.gameplay.map.map_objects.mobs.MobSpawn
import com.bapplications.maplemobile.gameplay.physics.Physics
import com.bapplications.maplemobile.gameplay.player.Player
import com.bapplications.maplemobile.input.EventsQueue.Companion.instance
import com.bapplications.maplemobile.input.InputAction
import com.bapplications.maplemobile.input.events.*
import com.bapplications.maplemobile.pkgnx.NXNode
import com.bapplications.maplemobile.utils.Point
import com.bapplications.maplemobile.utils.Rectangle
import com.bapplications.maplemobile.utils.StaticUtils
import kotlinx.coroutines.*

class GameMap(camera: Camera) : EventListener {
    var mapId = 0
        private set
    var state: State
        private set
    private var mobs: MapMobs? = null
    lateinit var player: Player
        private set
    val camera: Camera
    private var drops: MapDrops? = null
    private var mapInfo: MapInfo? = null
    private var physics: Physics? = null
    private var portals: MapPortals? = null
    private var tilesobjs: MapTilesObjs? = null
    private var characters: MapCharacters? = null
    private var backgrounds: MapBackgrounds? = null

    enum class State {
        INACTIVE, TRANSITION, ACTIVE
    }

    fun init(mapid: Int) {
        // drops.init();
        loadMap(mapid)
    }

    override fun onEventReceive(event: Event) {
        when (event.type) {
            EventType.ItemDropped -> {
                val (oid, id, start, owner, _mapId) = event as ItemDroppedEvent
                if (mapId == _mapId) {
                    spawnItemDrop(oid, id, start, owner)
                }
            }
            EventType.PickupItem -> {
                val (cid, oid, _mapId) = event as PickupItemEvent
                Log.e("GameMap", "Pick up item: $cid, $oid, $_mapId")
                val drop = drops!![oid] as Drop?
                drop?.let {
                    val char: MapObject? = if (cid == player.oid) {
                        player.pickupDrop(it)
                        player
                    } else {
                        characters?.getChar(cid)
                    }
                    if (mapId == _mapId) {
                        // pickup animation
                        char?.let { drops?.remove(oid, Drop.State.PICKEDUP, it) }
                    }
                }
            }
        }
    }

    private fun spawnMobs(src: NXNode) {
        var oid = 100 // todo: needs a way to calculate that
        for (spawnNode in src) {
            if (spawnNode.getChild<NXNode>("type").get("") != "m") {
                continue
            }
            val id = spawnNode.getChild<NXNode>("id").get("")
            val flip = spawnNode.getChild<NXNode>("f").get(0L).toShort()
            val p = Point(spawnNode)
            val spawn = MobSpawn(oid++, id.toInt(), 0.toByte(), 0.toByte(), flip, true, 0.toByte(), p.flipY())
            mobs!!.spawn(spawn)
        }
    }

    fun spawnItemDrop(oid: Int, id: Int, start: Point?, owner: Int) {
        val spawn = DropSpawn(oid, id, id == 0, owner, start!!, Drop.State.DROPPED, true)
        drops!!.spawn(spawn)
    }

    fun loadMap(mapid: Int) {
        mapId = mapid
        val strid = StaticUtils.extendId(mapid, 9)
        val prefix = strid[0]
        val src = if (mapid == -1) null else Loaded.getFile(Loaded.WzFileName.MAP).root.getChild<NXNode>("Map").getChild<NXNode>("Map$prefix").getChild<NXNode>("$strid.img")

        // in case of no map exist with this mapid
        if (src != null && !src.isNotExist) {
            try {
                mobs = MapMobs()
                drops = MapDrops()
                characters = MapCharacters()
                physics = Physics(src.getChild("foothold"))
                backgrounds = MapBackgrounds(src.getChild("back"))
                portals = MapPortals(src.getChild("portal"), mapid)
                mapInfo = MapInfo(src, physics!!.fht.walls, physics!!.fht.borders)
                tilesobjs = MapTilesObjs(src, Rectangle(mapInfo!!.walls, mapInfo!!.borders))
                spawnMobs(src.getChild("life"))
            } catch (e: Exception) {
                Log.e("GameMap", "Error loading map $mapid")
                throw e
            }
        }
    }

    fun respawn(position: Point?) {
        Music.play(mapInfo!!.bgm)

        val startpos = physics!!.getYBelow(position)
        player.respawn(startpos, mapInfo!!.isUnderwater)
        camera.setView(mapInfo!!.walls, mapInfo!!.borders)
        camera.update(player.position.negateSign())
    }

    fun update(deltatime: Int) {
        if (state != State.ACTIVE) return

        runBlocking {
////        combat.update();
            async {
                backgrounds!!.update(deltatime)
            }
            ////        effect.update();

            async {
                tilesobjs!!.update(deltatime)
            }
////        reactors.update(physics);
////        npcs.update(physics);
            async {
                characters!!.update(physics!!, deltatime)
            }
            async {
                drops!!.update(physics!!, deltatime)
            }
            async {
                mobs!!.update(physics, deltatime)
                player.update(physics!!, deltatime)
            }
            async {
                portals!!.update(player.position, deltatime)
            }
            async {
                camera.update(player.position)
            }
        }
        player?.stats?.canUseUpArrow?.postValue(portals?.collidePortal(player)!!
                || player.isClimbing || mapInfo!!.findLadder(player.position, true) != null)

        checkDrops()

        if (!player.isClimbing /* && !player.is_sitting()*/ && !player.isAttacking) {
            if (player.isPressed(InputAction.DOWN_ARROW_KEY)) checkLadders(false)
            else if (player.isPressed(InputAction.UP_ARROW_KEY)) checkLadders(true)
            if (player.isPressed(InputAction.UP_ARROW_KEY)) enterPortal()


//            if (player.isPressed(InputAction.SIT))
//            check_seats();

//            if (player.isPressed(InputAction.ATTACK))
//            combat.use_move(0);
        }
        if (player.isInvincible()) return
        val oid = mobs!!.findColliding(player)
        if (oid != 0) {
            val attack = mobs!!.createAttack(oid)
            if (attack.isValid) {
                val result = player.damage(attack)
            }
        }
    }

    private fun checkDrops() {
        val drop = drops!!.inRange(player)

        // button is set to visible but he didnt click yet
        if((drop != null || player.isPressed(InputAction.LOOT_KEY))
                != player.stats.canLoot.value) {
            player.stats.canLoot.postValue(drop != null)
        }
        // pressed pick up button now try to pick it
        if(drop != null && !drop.onPickProcess && !drop.isPicked() && player.isPressed(InputAction.LOOT_KEY)) {
            player.tryPickupDrop(drop)
        }
    }

    fun getPortalByName(portalName: String): Portal {
        return portals!!.getPortalByName(portalName)
    }

    private fun enterPortal() {
        if (player.isAttacking) return
        val warpinfo = portals!!.findWarpAt(player)
        if (warpinfo.intramap) {
            val spawnpoint = portals!!.getPortalByName(warpinfo.toname)
            val startpos = physics!!.getYBelow(spawnpoint.spawnPosition)
            player.respawn(startpos, mapInfo!!.isUnderwater)
        } else if (warpinfo.valid) {
            Sound(Sound.Name.PORTAL).play()
            GameEngine.instance?.changeMap(warpinfo.mapid, warpinfo.toname)
        }
    }

    fun draw(alpha: Float) {
        if (state != State.ACTIVE) return
        val viewpos = camera.position(alpha)
        backgrounds!!.drawBackgrounds(viewpos, alpha)
        for (id in Layer.values()) {
            tilesobjs!!.draw(id, camera.getPositionRect(), viewpos, alpha)
            physics!!.draw(viewpos)
//            reactors.draw(id, viewx, viewy, alpha);
//            npcs.draw(id, viewx, viewy, alpha);
            mobs!!.draw(id, viewpos, alpha)
            characters!!.draw(id, viewpos, alpha)
            player.draw(id, viewpos, alpha)
            drops!!.draw(id, viewpos, alpha)
        }
        //
//        combat.draw(viewx, viewy, alpha);
        portals!!.draw(viewpos, alpha)
        backgrounds!!.drawForegrounds(viewpos, alpha)
        //        effect.draw();

        // for debugging purposes
//        camera.draw();
    }

    fun clear() {
        state = State.INACTIVE
        instance.unregisterListener(EventType.ItemDropped, this)
        instance.unregisterListener(EventType.PickupItem, this)

        //        Texture.clear();
//        chars.clear();
//        npcs.clear();
//        mobs.clear();
//        drops.clear();
//        reactors.clear();
    }

    fun checkLadders(up: Boolean) {
        if (player.isClimbing || player.isAttacking) return
        player.ladder = mapInfo!!.findLadder(player.position, up)
    }

    fun enterMap(player: Player, portal: Portal) {
        this.player = player
        player.map = this
        state = State.ACTIVE
        respawn(portal.spawnPosition)
    }

    init {
        state = State.INACTIVE
        this.camera = camera
        instance.registerListener(EventType.ItemDropped, this)
        instance.registerListener(EventType.PickupItem, this)
    }
}